### ISSUE TITLE
Using managed disk as static volume seems to be sharing the disk between multiple deployments, even though each specifies a different disk URI

### ISSUE THREAD

**Issue by frachel** [Azure/AKS#2718]
Basically following the instructions here: https://docs.microsoft.com/en-us/azure/aks/azure-disk-volume I am running into the following problem. (the main difference is the example there uses a Pod type, whereas I am in a Deployment type, if that maters?

i created 2 managed disks. I have 2 deployments - they look like this (extraneous stuff stripped out)

```
apiVersion: apps/v1
kind: Deployment
metadata:
..
spec:
..
  template:
    ..
    spec:
      containers:
      - name: 50002-frank-rachel-devcontainer
        volumeMounts:
        - name: azure
          mountPath: /mnt/dev
      volumes:
      - name: azure
        azureDisk:
          kind: Managed
          diskName: AKSDisk
          diskURI: /subscriptions/ac38cd59-fdc9-4e2e-9d7a-1d18d3d35af3/resourceGroups/MC_dev-test-rg_dev-test-aks_eastus/providers/Microsoft.Compute/disks/frank-rachel-devcontainer-disk
```

And then a 2nd deployment, very similar but the diskURI is a completely different disk (and stuff like 50002-frank-rachel changes so they are unique)

I start up one of the deployments - connect to the pod (we have SSH running in the container, routed by a LoadBalancer service) and go to the /mnt/dev. I create a file "uday.txt".

I start up the second deployment.. connect.. the Linux prompt clearly shows the hostname being different so I know I'm in the 2nd container (ie, LoadBalancer routed me properly), and i see the "test1.txt" in the /mnt/dev.. 

(see later in the issue to see me connecting to both containers, showing different hostnames as expected, but sharing the disk)

When I describe the 2 pods I see they each show that they mounted a different disk URI:

```
Name:         50002-frank-rachel-devcontainer-99cc9c8c9-qz77p
Namespace:    50002-frank-rachel
Priority:     0
Node:         aks-pool1-34931576-vmss000000/10.6.45.4
Start Time:   Thu, 06 Jan 2022 13:14:56 -0500
Labels:       app=50002-frank-rachel-devcontainer
              pod-template-hash=99cc9c8c9
Annotations:  <none>
Status:       Running
IP:           10.6.45.15
IPs:
  IP:           10.6.45.15
Controlled By:  ReplicaSet/50002-frank-rachel-devcontainer-99cc9c8c9
Containers:
  50002-frank-rachel-devcontainer:
    Container ID:   containerd://9530b9ab02972836d868ab285169bf51e52080a294825b8e89da76c8cbe4bfee
    Image:          alipregistry.azurecr.io/alip/alip-developer:k8s-ssd
    Image ID:       alipregistry.azurecr.io/alip/alip-developer@sha256:d305bf25261c40ddc2ef5a2f70f99e22f05c0bd8fdfa74d42aa871b9f0a1aa17
    Port:           22/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Thu, 06 Jan 2022 13:15:14 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     4
      memory:  12Gi
    Requests:
      cpu:        1500m
      memory:     12Gi
    Environment:  <none>
    Mounts:
      /mnt/dev from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-pzn5z (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:         AzureDisk (an Azure Data Disk mount on the host and bind mount to the pod)
    DiskName:     AKSDisk
    DiskURI:      /subscriptions/ac38cd59-fdc9-4e2e-9d7a-1d18d3d35af3/resourceGroups/MC_dev-test-rg_dev-test-aks_eastus/providers/Microsoft.Compute/disks/frank-rachel-devcontainer-disk
    Kind:         Managed
    FSType:       ext4
    CachingMode:  ReadWrite
    ReadOnly:     false
  kube-api-access-pzn5z:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Normal   Scheduled               42m                default-scheduler        Successfully assigned 50002-frank-rachel/50002-frank-rachel-devcontainer-99cc9c8c9-qz77p to aks-pool1-34931576-vmss000000
  Warning  FailedMount             41m (x5 over 42m)  kubelet                  MountVolume.WaitForAttach failed for volume "disk.csi.azure.com-AKSDisk" : watch error:unknown (get volumeattachments.storage.k8s.io) for volume /subscriptions/ac38cd59-fdc9-4e2e-9d7a-1d18d3d35af3/resourceGroups/MC_dev-test-rg_dev-test-aks_eastus/providers/Microsoft.Compute/disks/frank-rachel-devcontainer-disk
  Normal   SuccessfulAttachVolume  41m                attachdetach-controller  AttachVolume.Attach succeeded for volume "disk.csi.azure.com-AKSDisk"
  Normal   Pulling                 41m                kubelet                  Pulling image "alipregistry.azurecr.io/alip/alip-developer:k8s-ssd"
  Normal   Pulled                  41m                kubelet                  Successfully pulled image "alipregistry.azurecr.io/alip/alip-developer:k8s-ssd" in 751.664876ms
  Normal   Created                 41m                kubelet                  Created container 50002-frank-rachel-devcontainer
  Normal   Started                 41m                kubelet                  Started container 50002-frank-rachel-devcontainer
```
And the second pod:

```
Name:         50001-uday-kher-devcontainer-7bccbc4b77-hk97v
Namespace:    50001-uday-kher
Priority:     0
Node:         aks-pool1-34931576-vmss000000/10.6.45.4
Start Time:   Thu, 06 Jan 2022 11:49:41 -0500
Labels:       app=50001-uday-kher-devcontainer
              pod-template-hash=7bccbc4b77
Annotations:  <none>
Status:       Running
IP:           10.6.45.14
IPs:
  IP:           10.6.45.14
Controlled By:  ReplicaSet/50001-uday-kher-devcontainer-7bccbc4b77
Containers:
  50001-uday-kher-devcontainer:
    Container ID:   containerd://f27f434196b0cee07d73a62f51cb3d2c99e4a9aa4ec63e8f4c977733a8c8dd29
    Image:          alipregistry.azurecr.io/alip/alip-developer:k8s-ssd
    Image ID:       alipregistry.azurecr.io/alip/alip-developer@sha256:d305bf25261c40ddc2ef5a2f70f99e22f05c0bd8fdfa74d42aa871b9f0a1aa17
    Port:           22/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Thu, 06 Jan 2022 11:49:42 -0500
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     4
      memory:  12Gi
    Requests:
      cpu:        1500m
      memory:     12Gi
    Environment:  <none>
    Mounts:
      /mnt/dev from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-9rdjr (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:         AzureDisk (an Azure Data Disk mount on the host and bind mount to the pod)
    DiskName:     AKSDisk
    DiskURI:      /subscriptions/ac38cd59-fdc9-4e2e-9d7a-1d18d3d35af3/resourcegroups/MC_dev-test-rg_dev-test-aks_eastus/providers/Microsoft.Compute/disks/uday-kher-devcontainer-disk
    Kind:         Managed
    FSType:       ext4
    CachingMode:  ReadWrite
    ReadOnly:     false
  kube-api-access-9rdjr:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  81m   default-scheduler  0/1 nodes are available: 1 Insufficient cpu, 1 Insufficient memory.
  Warning  FailedScheduling  81m   default-scheduler  0/1 nodes are available: 1 Insufficient cpu, 1 Insufficient memory.
  Normal   Scheduled         81m   default-scheduler  Successfully assigned 50001-uday-kher/50001-uday-kher-devcontainer-7bccbc4b77-hk97v to aks-pool1-34931576-vmss000000
  Normal   Pulling           81m   kubelet            Pulling image "alipregistry.azurecr.io/alip/alip-developer:k8s-ssd"
  Normal   Pulled            81m   kubelet            Successfully pulled image "alipregistry.azurecr.io/alip/alip-developer:k8s-ssd" in 408.71442ms
  Normal   Created           81m   kubelet            Created container 50001-uday-kher-devcontainer
  Normal   Started           81m   kubelet            Started container 50001-uday-kher-devcontainer
```

Here's what I see - when I ssh to the 1st container

```
dev@50001-uday-kher-devcontainer-7bccbc4b77-hk97v:~$ ls -ltr
total 4
-rw-r--r-- 1 dev dev 5 Jan  6 11:23 uday.txt
dev@50001-uday-kher-devcontainer-7bccbc4b77-hk97v:~$
```

You can see the prompt is "50001-uday...." which one of the deployments. The file "uday.txt" is there, 11:23 time.

I connect to the other container:

```
dev@50002-frank-rachel-devcontainer-99cc9c8c9-qz77p:~$ ls -ltr
total 4
-rw-r--r-- 1 dev dev 5 Jan  6 11:23 uday.txt
dev@50002-frank-rachel-devcontainer-99cc9c8c9-qz77p:~$
```

You can see I'm now in 50002-frank-rachel container.. but the same file is there.

From the "describes" above, it shows 2 disk URI's, but it seems they are sharing the disk.

Am I missing something, is this expected, or a bug?

Thanks.

**Comment by frachel** [comment]
I did just noticed the "FailedMount", but then a little after it says "SuccessfulAttachVolume" ?

**Comment by andyzhangx** [comment]
what's your aks version? and could you use the correct `DiskName` in the deployment? following `DiskName` is incorrect.
```
DiskName:     AKSDisk
```

**Comment by frachel** [comment]
> what's your aks version? and could you use the correct `DiskName` in the deployment? following `DiskName` is incorrect.
> 
> ```
> DiskName:     AKSDisk
> ```

Ugh, my bad. It was the DiskName - once I switched to my own naming convention after trying the sample I didn't update that to match the newly created disks. Changing that worked (I noticed it but wrongly assumed so long as the URI was correct it would work, and the name was just something to do self-referencing in the ymls).

Thanks, I'll close this.
